README.md
# async-methods (am)

> * Aids developer productivity by producing async processes within applications that are robust and with clear logic 
>
> * Eliminates indenting from complex async handlers making them easier to read and maintain
>
> * Supports **async functions /await**, **generators /yield** and **ES6 classes** for async operations
>
> * Extensible.  can add additional *Extended Promise* methods
>
> * am.js is one file with no dependencies, 235 unit and functional tests covering all use cases

Gitter now available for questions/support/suggestions on **async-methods** [gitter](https://gitter.im/ingenio-us-npm-packages/Lobby)

## Changes in version 1.0.0

Additional support added for async functions.  They can now be wrapped in same way as generators eg am(async function( .. ){ .. }) or am (async ( .. )=>{ .. })

async keyword can now be used in .next(fn), .error(fn), .mapFilter(fn), .map(fn), .filter(fn), .forEach(fn), .twoPrev(fn) and .threePrev(fn).  The await keyword can be used similarly to yield

Arrays and objects of async functions can be used in am.waterfall(), am.all(), am.parallel(), am.forEach()


##  README for version 0.1.x

[README-0.1.x.md](https://github.com/ingenious/async-methods/blob/master/README-0.1.x.md)

##  README for version 0.2.x

[README-0.2.0.md](https://github.com/ingenious/async-methods/blob/master/README-0.2.0.md)


## How it works

### Wrapping

am() can be used to wrap various types of entities such as generators, classes, promises, functions-with-callback etc. to generate an *Extended Promise*.  Every *Extended Promise* has the same set of chainable methods available to manipulate the resolved and/or rejected values
	
   *Chainable methods*

   - [**next**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#next)
   - [**error**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#error)
   - [**forEach**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#foreach)
   - [**map**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#map)
   - [**mapFilter**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#mapfilter)
   - [**filter**(&lt;fn | async fn| generator | (methodName,class)&gt;)](#filter)
   - [**twoPrev**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#twoprev)
   - [**threePrev**(&lt;fn | async fn | generator | (methodName,class)&gt;)](#threeprev)
   - [**prev**()](#prev)

   More: [.log()](#log), [.wait()](#wait), [.timeout()](#timeout), [.promise()](#promise) 

*Wrapping options*

   a. [am(&lt;**promise**&gt;)](#wrap-promises)  creates *ExtendedPromise* resolving to resolved value of input Promise or rejecting to rejected value of input Promise

   b. [am(&lt;**async function**&gt;)](#wrap-async-function)  creates *ExtendedPromise* resolving to returned value of async function or rejecting to any error occuring during the function.  Optional argunents can be added which are applied to the function when invoked. 

   
   c. [am(&lt;**generator**&gt;)](#wrap-generator)  creates *ExtendedPromise* resolving to returned value of input generator after any yield statements are resolved.  Any thrown errors values are rejected.  Optional argunents can be added which are applied to the generator when invoked. 
 
   d. [am(&lt;**function with callback argument**&gt;)](#wrap-function-with-callback)  creates *ExtendedPromise* resolving to returned value of callback or rejecting to error value in callback
    
   e. [am(**methodName**, &lt;**Class**&gt;)](#wrap-es6-class-with-methods) creates *ExtendedPromise* resolving to returned value of async method, generator method, or synchronous method, in anonymous or named class. Any thrown errors values are rejected to end of chain.   Optional argunents can be added which are applied to the function before the callback when invoked. 
   
   f. [am(**methodName**, new &lt;Class&gt;(arg))](#wrap-newed-class)  creates *ExtendedPromise* resolving to returned value of async method, generator method, or synchronous method, in new'ed anonymous or named class.  Any thrown errors values are rejected to end of chain
   
   g. [am(&lt;**Iterator**&gt;)](#wrap-iterator)  creates *ExtendedPromise* resolving to returned value of iterator (eg executed generator)
   
   h. [am(&lt;**boolean | number | string | array | object**&gt;)](#wrap-entities)  creates *ExtendedPromise* resolving to entity  
   
   i. [am.fn(&lt;**function without callback**&gt;,**args...**)](#wrap-no-callback-function)  creates *ExtendedPromise* resolving to returned value of function with args applied.  Any thrown errors values in function are rejected to end of chain
   
   j. [am.sfn(&lt;**function with success/fail callbacks**&gt;,args..)](#wrap-function-with-success-fail-callbacks)  creates *ExtendedPromise* resolving to returned value of success callback or rejecting to returned value of fail callback when provided arguments applied.

   k. [am(&lt;**Extended Promise**&gt;)](#wrap-extendedpromise)  creates identity
   
   l. [am.resolve(&lt;**entity**&gt;)](#wrap-resolve)  creates *Extended Promise* resolving to entity 
   
   m. [am.reject(&lt;**entity**&gt;)](#wrap-reject)  creates *Extended Promise* rejecting to entity 

More:    [am.all(&lt;**array or object of promises or generators**&gt;)](#wrap-all), [am.race(&lt;**array or object of functions-with-callback, promises or generators**&gt;)](#wrap-race), [am.forEach(&lt;**array or object of functions-with-callback, promises or generators**&gt;)](#wrap-foreach), [am.parallel(&lt;**array or object of functions-with-callback, promises or generators**&gt;)](#wrap-parallel), [am.waterfall(&lt;**array or object of functions-with-callback, promises or generators**&gt;)](#wrap-waterfall)

## Installation

[async-methods on npm](https://www.npmjs.com/package/async-methods)

See also  [api-responder on npm](https://www.npmjs.com/package/api-responder) [am-methods on npm](https://www.npmjs.com/package/am-methods)

In package.json

```javascript
                                                                                                                                                                                                   
	"async-methods":"^1.0.1"
	
```

In console

```
                                                                              
	$ npm install async-methods -P
  
```

In code

```javascript
                                                                                      
	let am=require('async-methods');
  
```

## Node versions and support for async/await

**async/await** is only available from version 7.6 of node.js.  If you are using an earlier version you will not be able to use the async/await features of **async-methods**.  **async-methods** will still work for wrapping generators and classes with normal functions and generator functions in versions earlier than 7.6.

Generators have been supported in nmodejs since at lease version 4.8.7

## Handling Promise rejections

NodeJS requires all rejections in a Promise chain to be handled or an exception is thrown.
When creating chains with **am** and *ExtendedPromise* methods always have an 

```javascript
	  .error(err=>{
	
	  }) 
``` 
or 

```javascript
	  .catch(err=>{
	
	  })  
```
	  
at the end of the chain (see examples below).  That way errors will be trapped and not cause exceptions
	
## Wrapping

### Wrap async function

#### am(async fn)  

Creates *ExtendedPromise* 


```javascript
                                                                                      
    
     am(async function(arg) {
       return await Promise.resolve({ a: 23864-arg })
     },4) // any extra arguments are applied to function
    .then(r => {
      console.log(r) //  { a: 23860 })
   
    })
    .error(err => {
       //handle error at end of chain
       console.log(err)
   
      })
    
  
```

### Wrap ES6 Class with methods

#### am( methodName , class { methodName { ... }}, args ...)

```javascript
                                                                                      
     am(56)
        .next('test', class {
            async test(value) {
              return await Promise.resolve(89 + (value || 0))
            }
          })
        
        .log()    //  145
        
        .error(err => {
			// handle errors at end of *ExtendedPromise* chain
          
        })
  
```

```javascript
                                                                
                                                                
let ep =  am(
        'test',
        class {
          async test() {
            return await Promise.reject({ e: 567 })
          }
        }
      )
        .next(r => {
        
        // Extended Promise rejects
          
        })
        .error(err => {
           console.log(err) // { e: 567 }
          
        })
    
``` 

### Wrap Newed Class

#### am(methodName, new class <name> {methodName (){}}(),args..)
                                                                                                                                
```javascript

    let ep = am(
        'test',
        new class {
          async test(value) {
            return await Promise.resolve(56 + (value || 0))
          }
        }(),
        54
      ).next(r => { 
      
        console.log(r) // 110
      }).error(err=>{
      
      		// handle errors at end of chain
      
      })

```

### Wrap Entities

#### am([3,4,5]) 

Create *ExtendedPromise* that returns an array.



```javascript
                                                                                      
   am([3, 4, 5]).mapFilter(function (value, i) {

     return 2 * value + i
   
   })
   .log('array wrapper') //  array wrapper [15ms] [ 6, 9, 12 ]​​​​​
   .error(err=>{
      		// handle errors at end of chain  
    
    })

```


```javascript
                                                                                      
   am([33, 4, 555])
   
   .wait(200)
   
   .filter(function* (value) {
   
     return yield am.resolve(4 - value);
   
   })
   
   .log('filter asyncronous,')  // ​​​​​filter asyncronous, [204ms] [ 33,  555 ]​​​​​
   
   .error(err=>{
      
      		// handle errors at end of chain
      
      })
  
```



##### Evaluate arrays and objects of async functions (or other types of asynchronous oeprations)

If the async function returns an array or object of async functions, Promises, functions with callback, generators thay are evaluated in the resolved response

```javascript
                                                                                      
   am(async function () {
      return {
        b: async function(){ return 'bb'},
        a: {
           b: async function(){ return 'bb'},
           a: {
              b: async function(){ return 'bb'},
              c: async function(){ return 'cc'},
             }
            }
        };
    })
    .log()   // { b: 'bb', a: { b: 'bb', a: { b: 'bb', c: 'cc' } } }​​​​​  
    .error(err=>{
      		// handle errors at end of chain
      
     })

  
  
```

### Wrap generator

#### am(generator)  

Creates *ExtendedPromise* (in same way as to 'co')

```javascript
                                                                                      
   am(function* () {
     
      return yield {
     
        b: Promise.resolve('bb'),
        a: {
           b: Promise.resolve('bb'),
           a: {
              b: Promise.resolve('bb'),
              c: Promise.resolve('cc')
             }
            }
        };
    
    })
    
    .log()
    
    // logs: 
    // yield object with async attributes { b: 'bb', a: { b: 'bb', a: { b: 'bb', c: 'cc' } } }​​​​​  
    
    .error(err=>{
      
      		// handle errors at end of chain
      
     })

  
  
```

#### Wrapping non-object

```javascript
                                                                                      
   am(4)
   .timeout(200)
   .filter(function* (value) {
   
      return yield am.resolve(4 - value);
   
   })
   
   .log('filter asyncronous non-object')  // filter asyncronous non-object [204ms]  null
   
   .error(err=>{
      
      		// handle errors at end of chain
      
      })
  
```

#### am({a:3}) 

Creates= *ExtendedPromise* that returns an object.

```javascript
                                                                                      
   am({ a: 34, b: 56, c: 78})
   
   .forEach(function (value, attr) {

       console.log(value, attr);// a 34 b 56 c 78

    })
    .error(err=>{
      
      		// handle errors at end of chain
      
    })
  
```

#### am(&lt;boolean | string | null&gt;)

Creates *ExtendedPromise* that returns entitity


```javascript
                                                                                      
   am(true).filter(function*(value){
     return value;
   })
   
   .log('other')  // 'other' true
   
   .error(err=>{
      
      		// handle errors at end of chain
      
   })
  
```


### Wrap Iterator

#### am(iterator)  

Creates *ExtendedPromise* which returns the result of the iterator 

```javascript
                                                                                      
   am(function*(a,b){
      a+=b;
      return yield a;
    }(45,55))
    
    .log('iterator')  // iterator 100
    
    .error(err=>{
      
      		// handle errors at end of chain
      
      })

```

### Wrap Function-with-callback

#### am(function(&lt;args&gt;, callback){ ... },&lt;args&gt;)

Creates *ExtendedPromise* that returns arguments of the callback and passes any err to a **.error(fn)**  or **.catch(fn)** at end of the chain.

```javascript
                                                                                      
   am(fs.readFile, __dirname + '/am.js')
  
     .then(function (content) {
     
        return content.toString().substr(-5);
     
     })
     
     .log('function with callback');   // function with callback '= am';​​​​​
     
     .error(err=>{
      
      		// handle errors at end of chain
      
      })
  
```

### Wrap Promises

#### am(&lt;Promise&gt;)

Creates *ExtendedPromise*

```javascript
                                                                                      
   am(Promise.resolve([45,67]))
    
      .map(function(item){
    
        return item/10;
    
   })
   .log('wrap promise')

   // logs
   // wrap promise  [ 4.5, 6.7 ]​​​​​
   
   .error(err=>{
      
      		// handle errors at end of chain
      
    })
  
```

### Wrap Function with Success-Fail callbacks

#### am.sfFn(function(<args>,successFn, errorFn,&lt;args&gt;)

returns Extended Promise that returns arguments of the success callback to **next()** or **then()** and passes the argument of the error function to a **.error(fn)**  or **.catch(fn)** at end of the chain.


```javascript
                                                                                      
  let sf = function (a, success, fail) {
    if (a < 10) {
        success(a);
    } else {
        fail('too big');
    }
  }
 

  am.sfFn(sf, 14)
   
   .log('sfFn','err')   // err 'too big'
   
   .error(err=>{      
      		// handle errors at end of chain
      
   })

  am.sfFn(sf, 1)
  .next(function (r) {
  
     console.log(r) // 1
  })
  .error(err=>{      
      		// handle errors at end of chain
      
  })
  
```

### Wrap No-callback Function

#### am.fn(fn [, args...])

Creates *ExtendedPromise* resolving to result of a function without a callback argument when the arguments provided are applied.  If error thrown in execution of function error value is rejected to end of chain

In some ways it is a convenience method as could also use am(fn(args..)), that is execute the function first begote wrapping
its main benefit is that it provides error trapping within the function

```javascript
                                                                          
     let noCallback=(a, b) => {
        return a + b
      },
     ep = am.fn(noCallback,345,678)
         .next(r => {
            console.log(r) // 1023
       
         })
         .error(err=>{      
      		// handle errors at end of chain
      
         })

```

Wrapping a non-async function with no callback provides arror hadling within an ExtendedPromise chain

```javascript
                                                                          
     let noCallback=(a, b) => {
       throw 'Something wrong'
        return a + b
      },
     ep = am.fn(noCallback,345,678)
         .next(r => {
            console.log(r) // no log
       
         })
         .error(err=>{      
          // handle errors at end of chain
          console.log(err) // Something wrong
      
         })

```

### Wrap *ExtendedPromise*

#### am(&lt;Extended Promise&gt;) 

Creates identity (input *ExtendedPromise*)

## Methods

> In all cases **fn** can be a **generator** or a normal function (for analagous synchronous operation)  An ES6 Class (anon or nameed) can be used using syntax .next(methodName,class).  This gives access to ***async/await***

An optional *tolerant* argument can be used with .map() or .filter() or with .mapFilter() to ensure completion even if there is an error


### map

#### .map(fn,tolerant)

*fn can be a normal function (synchronous operations) or a async funtion or generator ot class (asynchronous operations)*

equivalent to <array>.map().  If the previous stage of the chain resolves to an *array* or *object*, each element of the array or object is replaced with the returned result of the function or generator

#### map with function (synchronous step)

```javascript
                                                                                      
am(Promise.resolve([45, 67]))
  .map(function (item) {
    return item / 10;
  })
  
  .log('map Promise result') // map Promise result  [ 4.5, 6.7 ]​​​​​
  
  .error(err=>{      
      		// handle errors at end of chain
      
  })
  
```
####  map with async function and await

```javascript
                                                                                      
am([4, 5, 6])
  .map(async function(value) {
    return await Promise.resolve(2 * value)
  })
  .map(function(value) {
    return 3 * value
  })
  .log() //  [24, 30, 36]

  .error(err => {
    // handle errors at end of chain
  })
  
```
####  map with anonymous class and async/await

```javascript
                                                                                      
     am([4, 5, 6])
        .map(
          'asyncMap',
          class {
            async asyncMap(value) {
              return await Promise.resolve(2 * value)
            }
          }
        )
        .map(
          'syncMap',
          class {
            syncMap(value) {
              return 3 * value
            }
          }
        )
        .log()   //  [24, 30, 36]
        
        .error(err=>{      
      		// handle errors at end of chain
         })
  
```

### filter

#### .filter(fn, tolerant)

*fn can be a normal function (synchronous operations) or an async function or generator (asynchronous operations)   An ES6 Class (anon or named) can be used using syntax .filter(methodName,class).*

*Filter can be applied to objects and other entitites as well as arrays

#### filter with function (synchronous step)

```javascript
                                                                                      
  am(7).filter(function (value) {
    return 7 - value;
  })
  .log() // null
  .error(err=>{      
     // handle errors at end of chain
      
  })
  
```

####  filter with async function & await

```javascript
                                                                                      
am(7)
  .filter(async function(value) {
    return await (8 - value)
  })
  .log() // 7

  .error(err => {
    // handle errors at end of chain
  })
  
```

####  filter with generator/yield

```javascript
                                                                                      
  am(7).filter(function* (value) {
     return yield(8 - value);
  })
  
  .log() // 7
  
  .error(err=>{      
      		// handle errors at end of chain
      
  })
  
```

####  fiter with class

```javascript
                                                                                      
     am({ a: 4, b: 5, c: 6 })
        .filter(
          'asyncMap',
          class {
            async asyncMap(value, attr) {
              return await Promise.resolve(value < 5 ? false : true)
            }
          }
        )
        .filter(
          'syncMap',
          class {
            syncMap(value, attr) {
              return value === 6 ? false : 2 * value
            }
          }
        )
        
        .log()    // {b:5}
        
        .error(err=>{      
      		// handle errors at end of chain
      
        })
  
```

#### filter with object input

```javascript
                                                                                      
  am({
      a: 27,
      b: 78
  }).filter(function* (value, attr) {

    let a = yield Promise.resolve(value);
    return a > 50;
  })
  
  .log() //   { b: 78 }​​​​​
  
  .error(err=>{      
  		// handle errors at end of chain
      
  })
  
```

### mapFilter

#### .mapFilter(fn, tolerant)

Combines a map followed by a fiter using values returned from the map
If the mapping function for an element returns false, then the element is excluded from the result

*fn can be a normal function (synchronous operations) or an async function or generator (asynchronous operations).  An ES6 Class (anon or named) can be used using syntax .mapFilter(methodName,class).*


####  mapFilter with function (synchronous step)

```javascript
                                                                                      
     am([3, 4, 5])
       .mapFilter(function (value, i) {
         let a= 2 * value + i;
         return a > 6 ? a :false;
       })
       
       .log()     //   [ 9, 12 ]​​​​​
       
       .error(err=>{      
      		// handle errors at end of chain
      
       })
  
```

####  mapFilter with async function and await

```javascript
                                                                                      
am([4, 5, 6])
  .mapFilter(async function(value) {
    return value < 5 ? false : await Promise.resolve(2 * value)
  })
  .mapFilter(function(value) {
    return value === 10 ? false : 2 * value
  })
  .log() // [24]

  .error(err => {
    // handle errors at end of chain
  })
  
```

####  mapFilter with anonymous class

```javascript
                                                                                      
    am([4, 5, 6])
        .mapFilter('asyncMap', class {
            async asyncMap(value) {
              return value < 5 ? false : await Promise.resolve(2 * value)
            }
          })
        .mapFilter('syncMap', class {
            syncMap(value) {
              return value === 10 ? false : 2 * value
            }
          })
        .log()  // [24]
        
         .error(err=>{      
      		// handle errors at end of chain
      
         })
  
```

### forEach

#### .forEach(fn)

*fn can be a normal function (synchronous operations) or an async function or generator (asynchronous operations). An ES6 Class (anon or named) can be used using syntax .forEach(methodName,class).*

forEach returns an extended Promise resolving to the initial array or object

#### forEach with function (synchronous step)

```javascript
                                                                                      
    am([34, 56, 78]).forEach(function (value, i) {
           console.log(value, i);
    })
    
    .log()  // 34 0, 56 1, 78 2, [34, 56, 78]
    
    .error(err=>{      
      		// handle errors at end of chain
      
    })
  
```

#### forEach with async function / await (asynchronous steps)

```javascript
                                                                                      
  am([34, 56, 78])
  .forEach(async function(value, i) {
    console.log(await am.resolve(2 * value), i)
     // logs: 68 0, 112 1, 156 2
  })
  .error(err => {
    // handle errors at end of chain
  })
```
  
#### forEach with generator/yield (asynchronous steps)

```javascript
                                                                                      
  am([34, 56, 78]).forEach(function* (value, i) {
     console.log(yield am.resolve(2 * value),i);
     // 68 0, 112 1, 156 2
  })
  
  
  

changes.MD
# Changes to package by version

## version 1.0.0

1. setClass() allows a reference class to be set for a sequence of steps using class methods in a promise chain

2. clearClass() clears a reference class

3. async functions can now be wrapped in same way as generators eg am(async function( .. ){ .. }) or am (async ( .. )=>{ .. })

4.  async functions can now be used in .next(fn), .error(fn), .mapFilter(fn), .map(fn), .filter(fn), .forEach(fn), .twoPrev(fn) and .threePrev(fn).  The await keyword can be used similarly to yield

5.  Arrays and objects of async functions can be used in am.waterfall(), am.all()/am.parallel(), am.forEach()

### Fixes

Changed README docuemtnation for am.forEach(array || object) as previously incorrect

## version 0.2.15

1. **am.extend(extensionPackageList)**

async-methods can niow be extended progressively using **am.extend()** static mmthod.  Some new standard extensions are due for release soon, but applicastion developers can create their own and reference them by a filepath as well as by an npm module name.

```javascript
                                                                
    let am = require('am-mongo')

    am.extend(['../../am-axios.js', '../../am-cron.js'])

    // as well as additional am.xxx() static methods
    // am with now be extended with am-mongo, am-axios and am-cron methods available in the chain

```

This would have the same effect:

```javascript
                                                                
    let am = require('async-methods')

    am.extend(['am-mongo',../../am-axios.js', '../../am-cron.js'])

    // am with now be extended with am-mongo, am-axios and am-cron methods available in the chain
    // as well as additional am.xxx() static methods

```

2. ** Additonal arguments for class methods) When wrapping a class and specifying a method name, arguments to be passed to the method can be added can be added as arguments of the wrappimg am().  The same is now true for anonymous and named classes used as arguments to  

### next(methodName,class,...), error(methodName,class,...), 

  Additional arguments added to **next(methodName,class,...)** are **prepended** to the resolved result of previous step and appled as arguments to the method.  Thiis si useful if you don'tt want the result to be used by the method but wish to apply other arguments.

### twoPrev(methodName,class,...), threePrev(methodName,class,...)

  Additional arguments added to **twoPrev(methodName,class,...)** are **appended** to the resolved result of previous two steps and appled as arguments to the method.  The main purpose of twoPrev is to pass two results to a method.  If additional arguments are required they can be added in this way.

  Additional arguments added to **threePrev(methodName,class,...)** are **appended** to the resolved result of previous three steps and appled as arguments to the method.  The main purpose of threePrev is to pass three results to a method.  If additional arguments are required they can be added in this way.

3. Added tests for above feature

### Version 0.2.5

#### Changes to functionality

1. added .twoPrev(<generator|function|class>)

passes two arguments - last result and previous result in chain

returns Extended Promise resolving to an array of two latest resolved values in chain

2. added .threePrev(<generator|function|class>)

passes three arguments - last result, previous result and previous result in chain

returns Extended Promise resolving to an array of three latest resolved values in chain

#### Testing

3.  added tests for .twoPrev() and .threePrev()


### Version 0.2.4

#### Changes to functionality

1.  made class methods invokeable with a context (used in api-responder to pass state)

  eg 
  ```
               am.ExtendedPromise._applyResultToClass
                .apply(context, [argsHaveClass])
                .then(result => {
                  resolve(result)
                })
                .catch(reject)
```


### Version 0.2.0

#### Extended Promise Extensions enabled
1.  am._extend(ExtendedPromise) method added

2.  am-sample-extend.js module added as template

#### ES6 Classes with aysnc/await supported

3. Methods am.isClass(), am.argumentsHaveClass() utitlity added

4. am(method,class) detects and applies detects anonymous class with normal or async methods

5. am(initial) detects anonymous or named class and executes the constructor with any arguments



#### Testing

6. Test for am.race() (Generators) changed
7. mocha-wepack deprecated in favour of vanilla mocha
8. Tests added for am.isClass(), am.argumentsHaveClass()
9. Tests added for sample extension
10. Tests added for use of named and anonymous classes in next(), forEach(), map(),  mapFilter() and filter()


### Version 0.1.0

#### Changes to functionality

1. Changed behaviour of tolerant in map() to push passing items rather than assign to original index (same behaviour as  .mapFilter() )

2. Added ExtendedPromise property to am for testing of type

3. Changed behaviour of error() to a) mirror .catch when non function given as argument and b) when nothing returned from function (resolves to undefined)

4. Changed behaviour of next() to mirror .then() when non function given as argument - resolves to original value.


#### Fixes

5. Fixed second argument in .forEach and .filter (affects map, mapFilter) - i or attr values correctly passed

6. Corrected error in .forEach() documentation (resolved array is same as initial array)

7. Fixed issue with next() whereby an error in source Extended Promise was not passed through 

8. Fix for wrapping generator - any additional arguments now passed to generator function

#### Tests

9. Added test suite using mocha and webpack-mocha

10. Removed old tests

#### Re-factor

11. Removed ';' from statement ends

12. Internally, .next() and .error() used instead of .then() and .catch()

13. AM class renamed to ExtendedPromise




am-test.3.js
'use strict';

// .if(fn,fn) or .if(fn).next() .elseif(fn,fn) or .elseif(fn).next() .else(fn) or .else().next()  .while(fn,fn) or .while(fn).next()
// .do(fn) .. until(fn)  .interval(fn,n)  or interval(n)
// .extend()  .clone()  .repeat(fn,n)  .pickBy()
// .toJSON  .toObject .toString()?
// sync attribute : true|false?

let am = require('./am.js');



//==========================================================

//  Tests

//==========================================================

// console.log(am.isNextable([4, 5]));

//==========================================================

am.waterfall(
  [am.resolve(999).timeout(2000),
    am.resolve(8).timeout(1000)
  ]).log('waterfall');

am(function* () {
  return yield am(4);
}).log(483);

am([3, 4, 5]).mapFilter(function (value, i) {
  return 2 * value + i;
}).log(799);

am([3, 4, 5]).mapFilter(function (value, i) {
  let a = 2 * value + i;
  return a > 6 ? a : false;
}).log('mapfilter');

am([3, 4, 5]).map(function (value, i) {
  return 2 * value + i;
}).log(804);

am([3, 4, 5]).map(function* (value, i) {
  return yield am.resolve(3 * value + i);
}).log(820);

am.filter([34, 56], function* (value) {
  if (value === 34) {
    return yield true;
  } else {
    return yield Promise.reject(true);
  }
}, true).log('750 ok', '750 error');

am([34, 56, 78]).forEach(function (value, i) {
  console.log(value, i);
}).log();

am({
  a: 34,
  b: 56,
  c: 78
}).forEach(function (value, i) {
  console.log(value, i);
}).log();

am([34, 56, 78]).forEach(function* (value, i) {
  console.log(yield am.resolve(2 * value), i);
}).log();

am({
  a: 34,
  b: 56,
  c: 78
}).forEach(function* (value, attr) {
  console.log(73, yield am.resolve(3 * value), yield am.resolve(attr));
}).log('object async');

am(7).forEach(function* (value, i) {
  console.log(yield value, i);
}).log('non object');

am([3, 4, 5]).filter(function* (value) {
  return yield(4 - value);
}).log('filter syncronous, 799');

am([33, 4, 555]).timeout(200).filter(function* (value) {
  return yield am.resolve(4 - value);
}).log('filter asyncronous, 807');

am(4).timeout(200).filter(function* (value) {
  return yield am.resolve(4 - value);
}).log('filter asyncronous non-object');


am(7).filter(function (value) {
  return 7 - value;
}).log(810);

am(7).filter(function* (value) {
  return yield(8 - value);
}).log(813);

am.waterfall([7, function (a, cb) {
  cb(null, 444 + a);
}, function (a, b, cb) {

  cb(null, 777 + a + b);
}]).log('waterfall');

am.waterfall({
  a: am.reject(88),
  b: function (result, cb) {
    result.f = 567;
    cb(null, 444 + result.a);
  },
  c: function (result, cb) {
    cb(null, 444 + result.a + result.b);
  }
}).error(function (err) {
  console.log(701, err);
  return am.reject(new Error('no gogod'));
}).log('waterfall object', 'waterfall err');

am.resolve(89).timeout(2000).log('timeout');
am.race([am.resolve(999).timeout(2000), am.resolve(8).timeout(1000)]).log('race');

am.race(56).log('solo 56');

am.resolve(45).next(function (r) {
  console.log(this._state_);
  console.log(r);
  return 455;
}).next(function (r) {

  this._state_.prev.then(function (result) {
    console.log(result);
  });
  console.log(this._state_);
  console.log(r);

}).log('test');

am(function* () {
  return yield [
    Promise.resolve(90), [Promise.resolve(100), [Promise.resolve(100), Promise.resolve(190)]]
  ];
}).log('co ');

am([4, 5]).log('wrap array');

am(function* () {
  return yield {
    b: Promise.resolve('bb'),
    a: {
      b: Promise.resolve('bb'),
      a: {
        b: Promise.resolve('bb'),
        c: Promise.resolve('cc')
      }
    }
  };
}).log('yield object with async attributes');

am(function* () {
  return yield [
    Promise.resolve(90), Promise.resolve(190)
  ];
}).filter(function* (number) {
  return yield Promise.resolve(number < 100);
}).log('async filter');

// wrap a callback function
am(function (a, cb) {
  cb(null, a * 2);
}, 4).log();

am.resolve({
  a: 4
}).log('resolve');

am.reject(6).log('', 'Error:');
am.parallel([Promise.resolve(45), Promise.resolve(67)]).log('parallel');

am.all([Promise.resolve(89), Promise.resolve(8), [3, 4, Promise.resolve(43)]]).log();
am.all({
  a: 4,
  c: [Promise.resolve(89), Promise.resolve(8), [3, 4, Promise.resolve(43)]]
}).log();

am(function (cb) {
  cb(null, 4);
}).log('callback');

am.all(function (cb) {
  cb(null, 4);
}).log(559);

am.waterfall([function (cb) {
  cb(null, 4);
}, function (a, cb) {

  cb(null, 6 + a);
}, function (a, b, cb) {

  cb(null, 4 + b);
}, function (a, b, c, cb) {

  cb(null, 6 + a + b + c);
}]).log('560');

am.waterfall({
  aa: function (res, cb) {
    res.f = 67;
    cb(null, 4);
  },
  bb: function (res, cb) {
    cb(null, 6 + res.aa);
  },
  cc: function (res, cb) {
    cb(null, 4 + res.bb + res.aa);
  },
  dd: function (res, cb) {
    cb(null, 6);
  }
}).log(585);

am(function* (a, b) {


  a += b;
  return yield a;
}(45, 55)).log('iterator');
// iterator 100

// https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Iteration_protocols



am([2, 34].entries()).log('iterator II');

// iterator II undefined


am(true).filter(function* (value) {
  return yield value;
}).log('other');

// 'other' true


am(Promise.resolve([45, 67]))
  .map(function (item) {
    return item / 10;
  }).log('wrap promise');

// logs
// wrap promise  [ 4.5, 6.7 ]​​​​​

am.race({
  a: Promise.resolve(56),
  b: 45
}).log('race');

Promise.race([Promise.resolve(56), Promise.resolve(45)]).then(function (result) {
  console.log(result);
});